consensus/istanbul/backend/announce.go

```diff
@@ -516,59 +511,61 @@ func (sb *Backend) handleAnnounceMsg(peer consensus.Peer, payload []byte) error
 	shouldProcessAnnounce, err := sb.shouldGenerateAndProcessAnnounce()
 	if err != nil {
 		logger.Warn("Error in checking if should process announce", err)
-		return err
 	}
 
 	if shouldProcessAnnounce {
-		logger.Trace("Going to process an announce msg", "announce records", announceData.AnnounceRecords)
+		logger.Trace("Processing an announce message", "announce records", announceData.AnnounceRecords)
 		for _, announceRecord := range announceData.AnnounceRecords {
-			if announceRecord.DestAddress == sb.Address() {
-				enodeBytes, err := sb.decryptFn(accounts.Account{Address: sb.Address()}, announceRecord.EncryptedEnodeURL, nil, nil)
-				if err != nil {
-					sb.logger.Warn("Error decrypting endpoint", "err", err, "announceRecord.EncryptedEnodeURL", announceRecord.EncryptedEnodeURL)
-					return err
-				}
-				enodeURL := string(enodeBytes)
-				node, err := enode.ParseV4(enodeURL)
-				if err != nil {
-					logger.Error("Error parsing enodeURL", "enodeUrl", enodeURL)
-					return err
-				}
+			// Only process an announceRecord intended for this node
+			if announceRecord.DestAddress != sb.Address() {
+				continue
+			}
+			enodeBytes, err := sb.decryptFn(accounts.Account{Address: sb.Address()}, announceRecord.EncryptedEnodeURL, nil, nil)
+			if err != nil {
+				sb.logger.Warn("Error decrypting endpoint", "err", err, "announceRecord.EncryptedEnodeURL", announceRecord.EncryptedEnodeURL)
+				return err
+			}
+			enodeURL := string(enodeBytes)
+			node, err := enode.ParseV4(enodeURL)
+			if err != nil {
+				logger.Warn("Error parsing enodeURL", "enodeUrl", enodeURL)
+				return err
+			}
+			sb.lastAnnounceAnsweredMu.Lock()
+			// Don't answer an announce message that's been answered too recently
+			if lastAnswered, ok := sb.lastAnnounceAnswered[msg.Address]; !ok || time.Since(lastAnswered) < announceAnswerCooldownDuration {
```

Comment:
Why is the rate limit applied here instead of in `answerAnnounce`? It seems to be that we should always upsert, even if we don't send out a certificate. I would recommend moving the cooldown logic into `answerAnnounce`.

Its also a little unclear to me why we need a cooldown for this action because the response is to send an `enodeCertificate` (a fairly small message) directly back to the sender of this message, which does not risk flooding the network. If a cooldown is needed, 5 minutes seems extreme, especially because a number of scenarios can result in receiving an announce more frequently. I would suggest removing this cooldown or substantially reducing it (to something to 10 - 30 seconds)